feat(resharding): introduce resharding actor

Trisfald opened the pull request:
Main changes introduced by this PR:
- Calling `FlatStorageResharder` `resume` and `start` in all code paths.
- Added a very thin `ReshardingActor` and related Sender/Request items. Unfortunately, this meant propagating an extra argument through `Client` and `Chain`.
- Moved `FlatStorageResharder` in `ReshardingManager`. Wrapped inside `Option` because it makes initialization of `Chain` easier, and it's not needed in many tests anyway. 
- Refactored `FlatStorageResharder` to work with a `Sender`. Functionality is the same as before.
- A lot of uninteresting changes just to make old tests compile.

wacban commented on chain/chain/src/chain.rs:
```diff
@@ -361,7 +362,9 @@ impl Chain {
         let resharding_manager = ReshardingManager::new(
```
Do we even need resharding manager in the view client?

Trisfald replied:
I don't see a reason. I'll make a note to ask @shreyan-gupta and then remove the manager from there

shreyan-gupta replied:
Shouldn't be needed

wacban commented on chain/chain/src/flat_storage_resharder.rs:
```diff
@@ -52,13 +54,24 @@ use near_store::{ShardUId, StorageError};
 pub struct FlatStorageResharder {
     runtime: Arc<dyn RuntimeAdapter>,
     resharding_event: Arc<Mutex<Option<FlatStorageReshardingEventStatus>>>,
+    scheduler: messaging::Sender<FlatStorageSplitShardRequest>,
```
nit: Maybe import messaging? I don't think there is any ambiguity about it.

wacban commented on chain/chain/src/flat_storage_resharder.rs:
```diff
@@ -52,13 +54,24 @@ use near_store::{ShardUId, StorageError};
-    pub fn new(runtime: Arc<dyn RuntimeAdapter>) -> Self {
+    ///
+    /// # Args:
+    /// * `runtime`: runtime adapter
+    /// * `scheduler`: component used to schedule the background tasks
+    /// * `controller`: manages the execution of the background tasks
```
That's pretty!

wacban commented on chain/chain/src/flat_storage_resharder.rs:
```diff
@@ -254,8 +256,9 @@ fn retrieve_shard_flat_head(shard: ShardUId, store: &FlatStoreAdapter) -> Result
-fn split_shard_task(resharder: FlatStorageResharder, controller: FlatStorageResharderController) {
-    let task_status = split_shard_task_impl(resharder.clone(), controller.clone());
+pub fn split_shard_task(resharder: FlatStorageResharder) {
+    let task_status = split_shard_task_impl(resharder.clone());
```
Sanity check: What's up with the `clone`s?

wacban commented on chain/chain/src/flat_storage_resharder.rs:
```diff
@@ -501,7 +501,7 @@ pub enum FlatStorageReshardingTaskStatus {
 /// Helps control the flat storage resharder operation. More specifically,
 /// it has a way to know when the background task is done or to interrupt it.
-#[derive(Clone)]
+#[derive(Clone, Debug)]
```
I'm surprised that sender and receiver implement Debug but alright.

Trisfald replied:
Caught me by surprise as well

wacban commented on chain/chain/src/resharding/event_type.rs:
```diff
@@ -26,34 +26,35 @@ pub struct ReshardingSplitShardParams {
-    /// * `shard_layout`: the new shard layout
-    /// * `block_hash`: hash of the first block with `shard_layout`
+    /// * `next_shard_layout`: the new shard layout
+    /// * `block_hash`: hash of the last block with the shard layout before `next_shard_layout`
```
mini nit: `with the old shard layout` is less verbose and hopefully still clear what it means. Feel free to leave as is if you prefer the more precise one.

wacban commented on chain/chain/src/resharding/event_type.rs:
```diff
@@ -26,34 +26,35 @@ pub struct ReshardingSplitShardParams {
     /// * `prev_block_hash`: hash of the block preceding `block_hash`
     ///
-    /// Returns a [ReshardingEventType] if exactly one resharding change is contained in `shard_layout`, otherwise returns `None`.
+    /// Returns a [ReshardingEventType] if exactly one resharding change is contained in
+    /// `next_shard_layout`, otherwise returns `None`.
```
Is it actually possible for this method to return None? 

It sounds a bit like if there is non-1 number of events None would be returned but it actually returns an error right?

Trisfald replied:
The method will return none if there's no resharding at all in the shard layout, like when all children are child of themselves. It's tested in `parse_event_type_from_shard_layout`

wacban commented on chain/chain/src/resharding/manager.rs:
```diff
@@ -23,15 +27,27 @@ pub struct ReshardingManager {
     pub fn new(
         store: Store,
         epoch_manager: Arc<dyn EpochManagerAdapter>,
+        runtime_adapter: Arc<dyn RuntimeAdapter>,
```
Just for my education, why do you need the runtime adapter in this PR?

Trisfald replied:
The reason is that now `FlatStorageResharder` is owned and created by `ReshardingManager`. `runtime` is needed by `FlatStorageResharder`

wacban commented on chain/chain/src/resharding/manager.rs:
```diff
@@ -69,6 +85,15 @@ impl ReshardingManager {
+                ReshardingEventType::SplitShard(split_shard_event.clone()),
+                &next_shard_layout,
+            )?,
+            None => tracing::info!(target: "resharding", "flat storage resharder not initialized"),
```
should that be tracing::error?

wacban commented on chain/chain/src/resharding/resharding_actor.rs:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    pub fn handle_flat_storage_split_shard_request(&mut self, msg: FlatStorageSplitShardRequest) {
+        split_shard_task(msg.resharder);
```
nit: Would it be any nicer as a method?  `msg.resharder.split_shard_task()` ?

Trisfald replied:
Yes, I did the change and it looks nicer to me

wacban commented on chain/chain/src/resharding/types.rs:
```diff
@@ -0,0 +1,13 @@
+
+#[derive(actix::Message, Clone, Debug)]
+#[rtype(result = "()")]
+pub struct FlatStorageSplitShardRequest {
```
Can you add comments to the pub structs?

wacban commented on chain/chain/src/resharding/types.rs:
```diff
@@ -0,0 +1,13 @@
+
+#[derive(Clone, near_async::MultiSend, near_async::MultiSenderFrom)]
+pub struct ReshardingSender {
+    pub flat_storage_split_shard_send: Sender<FlatStorageSplitShardRequest>,
```
In the future we'd want a sender for any type of resharding request and SplitShard would be one of the variants in the Request enum. No need to do it now though.

wacban reviewed (approved):
LGTM

tayfunelmas commented on chain/chain/src/resharding/manager.rs:
```diff
@@ -80,6 +96,15 @@ impl ReshardingManager {
+                ReshardingEventType::SplitShard(split_shard_event.clone()),
+                &next_shard_layout,
+            )?,
+            None => tracing::error!(target: "resharding", "flat storage resharder not initialized"),
```
should not we panic if the resharder is not initialized? or it is based on some protocol versioning? if so, it should not error message right?

shreyan-gupta replied:
Converting resharding sender to non-option should remove this case.

tayfunelmas commented on chain/chain/src/flat_storage_resharder.rs:
```diff
@@ -609,7 +605,10 @@ mod tests {
 
     /// Generic test setup.
-    fn create_fs_resharder(shard_layout: ShardLayout) -> (Chain, FlatStorageResharder) {
+    fn create_fs_resharder(
```
nit. what does _fs_ stand for here? if flat_storage, can we expand the name?

tayfunelmas commented on chain/chain/src/resharding/manager.rs:
```diff
@@ -23,15 +27,27 @@ pub struct ReshardingManager {
     /// This typically happens when the main process is interrupted.
     pub resharding_handle: ReshardingHandle,
+    /// Takes care of performing resharding on the flat storage.
+    pub flat_storage_resharder: Option<FlatStorageResharder>,
```
I would prefer having this non-optional (maybe add a TODO as it is now convenient for this PR) in the future since as far as I know, the resharding will fail if this is not initialized yet? or can it still succeed?

shreyan-gupta replied:
Would too prefer to have this non-optional. I think we should address this as part of this PR itself. For the places where we don't have a sender or don't need to pass a sender, we can just pass in noop() sender.

Trisfald replied:
I'll try to remove the Option and pass a noop()!

shreyan-gupta commented on chain/client/src/test_utils/mock_resharding_sender.rs:
```diff
@@ -0,0 +1,16 @@
+    split_shard_request: Arc<RwLock<VecDeque<FlatStorageSplitShardRequest>>>,
+}
+
+impl CanSend<FlatStorageSplitShardRequest> for MockReshardingSender {
```
This is the old framework for tests, before test loop. Let's try to keep it as lean as possible and not introduce any new functionality here? It would be really really really really really really great it we can just have a noop sender and kill all this code and just say we don't support resharding for old framework.

Trisfald replied:
I'll try to kill this code and replace with the least amount of code just to make stuff compile

shreyan-gupta commented on chain/chain/src/resharding/resharding_actor.rs:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    pub fn handle_flat_storage_split_shard_request(&mut self, msg: FlatStorageSplitShardRequest) {
+        msg.resharder.split_shard_task();
```
I'm honestly not too happy that we are calling a function from resharder here.

For the purposes of reducing mental burden of identifying what happens where, ideally all the logic to deal with the heavy task of splitting should have been a part of the ReshardingActor class. This is too much of indirection to deal with.

Is it possible for us to push the logic of splitting the flat storage here?

Trisfald replied:
I thought the general sentiment was to keep the actor as simple as possible, thus all the work and unit testing is done in the resharder. Any other thoughts?

shreyan-gupta replied:
Would love input from other folks, @wacban @Longarithm!

wacban replied:
I don't have strong feelings either way. My only recommendation would be to keep the actor api part (typically the handle method) as slim as possible. This way we can properly unit test the actual implementation without needed to spin up the actor or worry about async. So something like this:

```
fn handle(request) {
  the_actual_implementation(request); 
} 

mod tests {
  // here the the_actual_implementation can be unit tested without needing to bootstrap an entire actor
```

shreyan-gupta commented on nearcore/src/lib.rs:
```diff
@@ -401,6 +405,7 @@ pub fn start_with_config_and_synchronization(
         partial_witness_actor.clone().with_auto_span_context().into_multi_sender(),
         true,
         None,
+        resharding_sender.into_multi_sender(),
```
Ugh, we need a better way to handle adding new actors and new actix requests. Note to self; I'll see if I can come up with better ways of handling all actor senders in a single unified struct.

shreyan-gupta commented on chain/chain/src/resharding/manager.rs:
```diff
@@ -23,15 +27,27 @@ pub struct ReshardingManager {
+            FlatStorageResharder::new(
+                runtime_adapter,
+                sender.into_sender(),
+                FlatStorageResharderController::from_resharding_handle(resharding_handle.clone()),
```
I'm still unsure how does the controller work and what do we do with the sender and receiver here? I'll have to go back and do a quick read up on how crossbeam_channels work but we should ideally not be introducing a new framework here given we have actors and message senders? Thoughts?

Trisfald replied:
I planned to remove the channels in another PR because I thought they could come in handy but they did not. I can push the change here

shreyan-gupta reviewed (commented):
Looks good overall but have a couple of comments

Trisfald commented on chain/chain/src/flat_storage_resharder.rs:
```diff
@@ -195,20 +195,12 @@ impl FlatStorageResharder {
         let resharder = self.clone();
-        let task = Box::new(move || split_shard_task(resharder, controller));
-        scheduler.schedule(task);
+        self.scheduler.send(FlatStorageSplitShardRequest { resharder });
```
Yes, it should be easy to make the actor hold copy of resharder

Trisfald commented:
I've addressed most of the comments!

I left out two things:
- Making the `ReshardingActor` hold a resharder instance: it's a bigger change than expected because it introduces a circular dependency between the two
- Merging `ReshardingActor` with `FlatStorageResharder`: resharder is used both by the actor and by the logic which handles flat state status at startup, so it is convenient to have this separation at the moment

I plan to do further refactoring and improvements in separate PRs
